### PR TITLE
Refactor email manager

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR refactors the email manager to reduce the amount of repeated lines of code.

Fixes # (issue)

**Environment**: Staging(main branch) / Production

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Chore (refactoring code, technical debt, workflow improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- [ ] Test A
- [ ] Test B

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't performed a self-review of my own code and corrected any misspellings
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/emails/email-manager.ts [hunk 1]
```diff
@@ -27,45 +27,32 @@ import OrganizerScheduledEmail from "./templates/organizer-scheduled-email";
...
+  return new Promise((resolve, reject) => {
+    try {
+      const email = prepare();
+      resolve(email.sendEmail());
+    } catch (e) {
+      reject(console.error(`${prepare.constructor.name}.sendEmail failed`, e));
+    }
+  });
+};
+
```
Comment (sean-brydon) [inline packages/emails/email-manager.ts:41]:
```suggestion
interface EmailPreparation<T extends BaseEmail> {
  new (): T;
}

function sendEmail<T extends BaseEmail>(prepare: () => EmailPreparation<T>): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      const EmailClass = prepare();
      const email = new EmailClass();
      email.sendEmail();
      resolve();
    } catch (e) {
      reject(console.error(`${prepare.constructor.name}.sendEmail failed`, e));
    }
  });
}
```

I manage to make some improvements trying to type this correctly but didnt get much further than this. 

@zomars can your TS wizardry help me out here

Comment (sean-brydon) [reply packages/emails/email-manager.ts:41]: I actually managed to get this to work... Not sure if its the prettiest solution there is.
CleanShot 2023-02-22 at 03 56 02@2x

Comment (joeauyeung) [reply packages/emails/email-manager.ts:41]: Thanks for finding this @sean-brydon. I'll wait and see what @zomars thinks before merging.

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -1,7 +1,6 @@
 name: PR Update
 
 on:
-  push:
```
Comment (hariombalhara) [inline .github/workflows/pr.yml:4]: Is this intentional? @joeauyeung

File: packages/emails/email-manager.ts [hunk 3]
```diff
@@ -27,45 +27,32 @@ import OrganizerScheduledEmail from "./templates/organizer-scheduled-email";
 import type { TeamInvite } from "./templates/team-invite-email";
 import TeamInviteEmail from "./templates/team-invite-email";
 
+// eslint-disable-next-line
+const sendEmail = (prepare: () => any) => {
```
Comment (hariombalhara) [inline packages/emails/email-manager.ts:31]:
```suggestion
const sendEmail = (prepare: () => BaseEmail) => {
```

This seems to compile correctly.

File: packages/emails/email-manager.ts (hunk 1 above)
Comment (hariombalhara) [reply packages/emails/email-manager.ts:41]: Maybe this is acceptable https://github.com/calcom/cal.com/pull/7270/files#r1115417367

File: packages/emails/email-manager.ts [hunk 4]
```diff
@@ -27,45 +27,32 @@ import OrganizerScheduledEmail from "./templates/organizer-scheduled-email";
 import type { TeamInvite } from "./templates/team-invite-email";
 import TeamInviteEmail from "./templates/team-invite-email";
 
+// eslint-disable-next-line
```
Comment (hariombalhara) [inline packages/emails/email-manager.ts:30]:
```suggestion
```

File: packages/emails/email-manager.ts [hunk 5]
```diff
@@ -27,45 +27,32 @@ import OrganizerScheduledEmail from "./templates/organizer-scheduled-email";
 import type { TeamInvite } from "./templates/team-invite-email";
 import TeamInviteEmail from "./templates/team-invite-email";
```
Comment (hariombalhara) [inline packages/emails/email-manager.ts:28]:
```suggestion
import TeamInviteEmail from "./templates/team-invite-email";
import type BaseEmail from "./templates/_base-email";
```

**Comment by hariombalhara** [comment]
Merging this to the base branch. We can do a TS specific iteration later.

**Pull request merged**